python/lsst/daf/butler/core/datasets/type.py

```diff
@@ -532,6 +532,35 @@ def makeAllComponentDatasetTypes(self) -> List[DatasetType]:
             for componentName in self.storageClass.allComponents()
         ]
 
+    def overrideStorageClass(self, storageClass: str | StorageClass) -> DatasetType:
+        """Create a new `DatasetType` from this one but with an updated
+        `StorageClass`.
+
+        Parameters
+        ----------
+        storageClass : `str` or `StorageClass`
+            The new storage class.
+
+        Returns
+        -------
+        modified : `DatasetType`
+            A new dataset type that is the same as the current one but with
+            a different storage class.
+
+        Notes
+        -----
+        If this is a component dataset type, the parent storage class
+        will be retained.
+        """
```

Comment:
We could. That would force the storage class and python type to be loaded early. That might be a good idea, although that would lead to different error reporting if someone passes in an override vs someone passing in an externally-created dataset type (which is how pipelines work). This override will trigger the incompatibility error later on. I'm a bit torn -- raising early is generally better than raising late (and the stack trace from formatters is a nightmare until python 3.11 lets us easily add an extra context message to the existing stack trace). I think we would have to be a bit looser on the compatibility check here because we don't know if this is an override for write or override for read so would only complain if it's not compatible with either direction, whereas the later test is explicitly for reading. I think I've talked myself into doing the basic test.